File: src/main/java/com/example/estantevirtual/controller/BookController.java
package com.example.estantevirtual.controller;

import com.example.estantevirtual.exception.ResourceNotFoundException;
import com.example.estantevirtual.model.Book;
import com.example.estantevirtual.model.Options;
import com.example.estantevirtual.service.BookService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.Optional;
import java.util.UUID;

@RestController("/v1")
public class BookController {

    @Autowired
    BookService bookService;

    @GetMapping("/books/{id}")
    public ResponseEntity<?> findBook(@PathVariable UUID id) {
        Optional<Book> book = bookService.findBookBy(id);
        if (book.isEmpty()) {
            throw new ResourceNotFoundException("Recurso não encontrado");
        }
        return new ResponseEntity<>(book.get(), HttpStatus.OK);
    }

    @GetMapping("/books")
    public ResponseEntity<?> findBooks(
            @RequestParam(value = "page", defaultValue = "0", required = false) int page,
            @RequestParam(value = "page_size", defaultValue = "10", required = false) int pageSize,
            @RequestParam(value = "order_by", required = false) String oderBy,
            @RequestParam(value = "sort", defaultValue = "ASC", required = false) String sort
    ) {
        Options options = new Options(page, pageSize, oderBy, sort);
        Page<Book> books = bookService.findBooks(options);
        return new ResponseEntity<>(
                bookService.responseBuilder(books),
                HttpStatus.OK
        );
    }

    @PostMapping("/books")
    public ResponseEntity<?> saveBook(@Valid @RequestBody Book book) {
        bookService.saveBook(book);
        return new ResponseEntity<>(HttpStatus.OK);
    }

    @DeleteMapping("/books/{id}")
    public ResponseEntity<?> deleteBook(@PathVariable UUID id) {
        if (bookService.deleteBook(id)) {
            return new ResponseEntity<>(HttpStatus.OK);
        }
        throw new ResourceNotFoundException("Recurso não encontrado");
    }

    @PutMapping("/books/{id}")
    public ResponseEntity<?> updateBook(@PathVariable UUID id, @RequestBody Book book) {
        if (bookService.updateBook(id, book)) {
            return new ResponseEntity<>(HttpStatus.OK);
        } else {
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }
    }
}

File: src/main/kotlin/com/estantevirtual/controller/BookController.kt
package com.estantevirtual.controller

import com.estantevirtual.exception.ResourceNotFoundException
import com.estantevirtual.model.Book
import com.estantevirtual.model.Books
import com.estantevirtual.model.Options
import com.estantevirtual.service.BookService
import org.slf4j.Logger
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*
import javax.validation.Valid

@RestController
@RequestMapping("/v1/books")
class BookController(
    private val bookService: BookService,
    private val logger: Logger
) {
    @GetMapping("/{isbn}")
    fun findBook(@PathVariable isbn: String?): ResponseEntity<*> {
        val book = bookService.findBookByIsbn(isbn)?.orElse(null)
        book ?: run {
            logger.error("Not found isbn: $isbn")
            throw ResourceNotFoundException()
        }
        return ResponseEntity(book, HttpStatus.OK)
    }

    @GetMapping
    fun findBooks(
        @RequestParam(value = "page", defaultValue = "0", required = false) page: Int,
        @RequestParam(value = "per_page", defaultValue = "10", required = false) perPage: Int,
        @RequestParam(value = "order_by", defaultValue = "id", required = false) orderBy: String?,
        @RequestParam(value = "sort", defaultValue = "ASC", required = false) sort: String?
    ): ResponseEntity<Books> {
        val books = bookService.findBooks(Options(page, perPage, orderBy, sort))
        return ResponseEntity(books, HttpStatus.OK)
    }

    @PostMapping
    fun saveBook(@RequestBody book: @Valid Book?): ResponseEntity<*> {
        bookService.saveBook(book)
        return ResponseEntity<Any>(book, HttpStatus.CREATED)
    }

    @PutMapping("/{isbn}")
    fun updateBook(@PathVariable isbn: String, @RequestBody book: Book?): ResponseEntity<*> {
        return if (bookService.updateBook(isbn, book)) {
            book?.isbn = isbn
            ResponseEntity(book, HttpStatus.OK)
        } else {
            book?.isbn = isbn
            bookService.saveBook(book)
            ResponseEntity(book, HttpStatus.CREATED)
        }
    }

    @DeleteMapping("/{isbn}")
    fun deleteBook(@PathVariable isbn: String?): ResponseEntity<*> {
        return if (bookService.deleteBook(isbn)) {
            ResponseEntity<Any>(HttpStatus.NO_CONTENT)
        } else {
            logger.error("Not found isbn: $isbn")
            throw ResourceNotFoundException()
        }
    }
}
